### PR TITLE
CUDA support in Libxc

### PR DESCRIPTION
**Pull request by mfherbst**
Upstream Libxc has built-in CUDA support, which would be lovely to have in Yggdrasil as well to jump-start our GPU efforts in DFTK. I've seen something similar for SuiteSparse by @maleadt and @ViralBShah so I thought to give it a try. Let's see how far we get :).

### CODE REVIEW

**Comment by ViralBShah** [comment]
Nice!

**Comment by mfherbst** [comment]
Note: This succeeded, but actually did not yet build the CUDA version. Seems like CUDA support in the latest libxc is broken.

File: L/Libxc/Libxc/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    -DENABLE_FORTRAN=OFF \
+    -DDISABLE_VXC=OFF -DDISABLE_FXC=OFF -DDISABLE_KXC=ON -DDISABLE_LXC=ON ..
+
+make -j${nproc}
+make install
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms(; experimental=true)
```
Comment (giordano) [inline L/Libxc/Libxc/build_tarballs.jl:24]:
```suggestion
platforms = supported_platforms()
```

**Comment by mfherbst** [comment]
Ok, seems to have built the library with CUDA support. Can I easily test the produced binaries or shall we merge it and then I try directly on the produced wrappers?

**Comment by giordano** [comment]
https://docs.binarybuilder.org/stable/FAQ/#Can-I-publish-a-JLL-package-locally-without-going-through-Yggdrasil?-1

**Comment by mfherbst** [comment]
@giordano Sorry about the delays. I finally managed to properly do the testing and I'm sure this does the right thing. So this is ready for merge from my point of view.

**Comment by giordano** [comment]
Shouldn't this be a single package which can use CUDA as platform tag?

**Comment by mfherbst** [comment]
I basically copied what had been done with SuiteSparse at the time. If there is a better way to solve this, I'm all ears. And I'm afraid I don't fully understand your comment @giordano.

**Comment by maleadt** [comment]
There shouldn't be a separate package, just separate artifacts which will get loaded depending on whether there's CUDA support or not. That said, I haven't personally tested this mixed case; we currently set the `cuda` tag to `none` when the platform augmentation detects no CUDA toolkit, so I guess it should work if the artifact doesn't declare anything for the `cuda` tag?

**Comment by mfherbst** [comment]
I'm not sure that cuts it, because the way Libxc handles CUDA support is that it internally compiles different code paths (but exposed under the same symbol) depending on whether CUDA is enabled or not.

Effectively the library with and without CUDA are thus two different binary artefacts. The way I view it downstream then has to decide which binary to use depending on whether data sits on the CPU or on the GPU.

**Comment by maleadt** [comment]
> The way I view it downstream then has to decide which binary to use depending on whether data sits on the CPU or on the GPU.

You don't need a separate JLL for that. The augmentation takes care of only selecting the CUDA-built artifact if CUDA is available.

**Comment by mfherbst** [comment]
I see, but I'm still not sure this is what should be done here.

Because if I get it right when CUDA is available in your proposal I could only use Libxc with data that's on the GPU (since the CPU version of the functions is not compiled into the CUDA-enabled binary). While this is ok for my current use case I could imagine scenarios in the future where one want to mix running computations on GPU and CPU in an application and thus one would want to be able to dynamically dispatch to the GPU or the CPU versions of Libxc (i.e. essentially hook into two binaries from the package).

I'm ok to say we solve this in the future and for now using CUDA means Libxc cannot be used with the CPU, but this could be counterintuitive, so I wanted to make sure this limitation is clear.

**Comment by maleadt** [comment]
> the CPU version of the Libxc functions are not compiled into the CUDA-enabled binary

Ah yes, in that case you want multiple JLLs.

**Comment by mfherbst** [comment]
@maleadt @giordano Thanks for the discussion and your help!

**Pull request merged**